### DISCUSSION TITLE
Modeling go / no-go tasks

### DISCUSSION THREAD

**Discussion by andrewheusser** [mwshinn/PyDDM#44] [Ask for help]
Very cool package! Apologies if I missed this in the documentation. Does PyDDM support the ability to model go no-go task data? HDDM supports this functionality following the procedure outlined here by [Ratcliff et al 2018](https://cpb-us-w2.wpmucdn.com/u.osu.edu/dist/6/60429/files/2018/07/decision18-2kwb3qz.pdf). Very interested in the ability to do GDDM high throughput model eval, but I'm stuck with go no-go data. Would definitely consider a PR if you think its possible / could provide a little guidance on implementation.

**Comment by mwshinn** [comment]
Glad to hear you are interested in using PyDDM for go-nogo tasks!

Yes, it is possible to do one-choice RTs (e.g. go-nogo paradigm). There are a few different ways to do it.  The Ratcliff one, if I understand correctly, just assumes nogo if you hit the bottom boundary.  But if you don't need collapsing bounds, you can have a (near-) infinite bound on the bottom if you set the bounds to be relatively large compared to the drift rate, and then set the starting position to be close to the top bound.  You can get arbitrarily small bound crossings on the lower boundary this way.  

Whether or not you want infinite bounds on the bottom, you probably will want to fit using a loss function that only takes the upper bound into consideration.  By default, fitting using likelihood in PyDDM considers both the upper and the lower bound.  You can do this by writing a custom loss function that doesn't take "error rate" (i.e. the lower bound) into consideration when determining likelihood.  See:

https://pyddm.readthedocs.io/en/latest/cookbook/loss.html

for info on creating a custom loss function.  (Note that you currently probably don't want to use the "undecided" feature with the default loss functions - this isn't too well supported right now and should probably be removed.)

Also, since this seems like it would be useful for many people, I'd be happy to add any code you write to the official documentation in the cookbook - just let me know!

Hope this helps!

**Reply by andrewheusser** [reply]
Thanks for the quick response, Maxwell! Maybe we are saying the same thing, but I'm thinking about modeling the go no-go task as a 2-choice task where the go condition is explicit and the no-go condition is implicit / lacks a motor response. In a model with a near infinite lower bound, we'd expect 0 or very few lower bound crossings, but this isn't necessarily what I'm after. In theory, I'd like to leverage the RTs + accuracy for the go condition and just accuracy for the no-go condition.

I am definitely interested in testing models with collapsing bounds, so some solution robust to boundary functions is desirable.

The HDDM [solution](https://github.com/hddm-devs/hddm/blob/master/hddm/likelihoods.py#L138-L149) appears to be separately modeling trials with responses (using a wiener distribution) and no response trials (using a binomial distribution) and summing them to get an overall model log likelihood. 

I'm just not familiar enough with the Fokker-Planck equation to understand whether/how something like this would work. Hope this makes some sense. Perhaps a custom loss function that separately models and incorporates the no-go trials makes sense here?

I'd be happy to contribute some code/example recipes if I can figure out that the approach looks like!

**Reply by mwshinn** [reply]
Oh, I see.  If you don't need to model a near-infinite lower bound, then this is really easy to do on PyDDM.  All you need to do is create a custom loss function.  The terminology will be a bit different: PyDDM uses the term "correct" for upper crossings and "incorrect" for lower crossings.  So in this model, "correct" would be "go" and "incorrect" would be explicit nogo.  PyDDM would give you an RT distribution for correct ("go") and for error ("nogo").

So basically you will have three types of trials: upper bound crosses, lower bound crosses, and non-bound crosses.  The latter two are equivalent and do not depend on RT, so you could just ignore "error RTs".  I don't fully understand why the method used in HDDM leads to a likelihood and fixes the problem of mixing discrete with continuous likelihoods, but if my understanding of the way they handle nogos is correct, it doesn't seem like it will translate from DDMs to GDDMs.  So you might need to figure out a new way of doing this.  If I were going to do it, I might use a "pseudo-likelihood" by combining the likelihood for upper bound crossings with a binomial distribution for choice with weights depending on how important they are to your fit.  Then you can, e.g., add a separate weight to distinguish between hits and correct rejection as well as between misses and false alarms.  It won't be a proper likelihood, but it will work.

So in summary, the main thing you need to do is create a custom loss function.  See the link I posted above.  The way to calculate the likelihood for "correct" will be similar/identical LossLikelihood, but you'll need to do something different for the others.  I'd suggest ignoring the incorrect distribution and taking 1-sum(correct) because this will also handle particles which didn't cross either bound.  You could also consider non-likelihood based loss functions - again, see the link above for examples of those.

**Reply by andrewheusser** [reply]
Thank you, this is very helpful. I'll play around with some candidate loss functions and see what I can come up with. At this point, it seems like I may be able to do this without modifying the package but will keep you in the loop if I need to change anything (and will submit a PR if it seems useful to others)
